### ISSUE TITLE
internal service errors

### ISSUE THREAD

**Issue by lukeolson** [inducer/relate#116]
Still getting a number of service errors on the latest change:
```
OperationalError: could not serialize access due to read/write dependencies among transactions
DETAIL:  Reason code: Canceled on identification as a pivot, during conflict in checking.
HINT:  The transaction might succeed if retried.
```

This happened today with approximately 150 using an inclass flow.  I wonder if the overhead is too high as participants had trouble connecting to the server without delay.

**Comment by inducer** [comment]
First of all, sorry you had trouble! (And for anyone else following, we've enabled the `SERIALIZABLE` transaction isolation level for Relate on the UIUC instance, which is causing the error quoted above, for the benefit that things like `if_has_fewer_sessions_than` can now be strictly enforced even in the presence of "racing" requests.) As I said on the monitoring list, I think I've got a plan for this particular (needless) bottleneck--all the errors we saw today are for GET requests to flow pages. Since those (mostly) don't update state, they have low demands on consistency and really shouldn't be wrapped in (expensive!) transactions.

That said, I'm happy to end the `SERIALIZABLE` experiment if you decide it's not worth it. Just know that without that, we'll need to accept occasional consistency glitches because of data races. To me, what's happening is what I meant when I said that enabling `SERIALIZABLE` would temporarily increase the error rate--we just need to see what transactions race, fix the ones that are fixable, and retry the rest. The win sitting opposite that (temporary, hopefully) loss is a better system at the end of the day.

Let me know what you think.

**Comment by inducer** [comment]
The fix for the errors above is in 24bdb12693ee3dc24ca42be324574c71c9aab600 and is now live. I of course can't promise that this is the last type of transaction race we'll encounter, but at least this one should be resolved. I'll close this, and we'll see what else we get.
